### PR TITLE
Updating order cycle will ensure any $NaN subscription item prices to be recalculated

### LINKED ISSUE #9100: $NaN in subscriptions back office
## Description

The admin for enterprise ID#3651 created a new order cycle with 2 products that were appearing as $NaN in the subscriptions back office.

Customer support confirmed that subscription orders had been correctly placed and processed by the system. However, it is a confusing back office experience and wastes enterprises time to check that orders are placed and processed correctly, especially when the workaround of creating a new order cycle does not alleviate the problem.

## Expected Behavior

@RachL mentioned in [this thread](https://openfoodnetwork.slack.com/archives/CG7NJ966B/p1615379974016200?thread_ts=1615373982.015700&cid=CG7NJ966B) that the "$NaN" sometimes appears in the subscription back office when there is no upcoming order cycle that includes the relevant products. This is a display issue but does not impact order placement or processing. The workaround she suggested was that if you open an order cycle that contains the product or set it to open in the future, then the $NaN goes away. 

## Actual Behaviour

However, when one of our hub admin's created a new order cycle to fix this, the $NaN did not go away. For all subscriptions where variants were added at the same time, all price estimates are displayed in the subscription back office as well as in Metabase. For the 2 specific subscriptions where 2 product variants were added to each subscription at a date after the subscription was created, the price estimate is not displayed in Metabase and appears as $NaN in the back office even after an order cycle is opened. Refer screenshots.

## Steps to Reproduce
Order cycle name "CYCLE 10: 15 April 22 (Week 15)" (OC ID #12116)

Product info
Product name: Full cream milk, 2 litre
Product ID: #15982
Variant ID: #40925
Supplier Name: Saint David Dairy
Supplier ID: # 826
Created at: August 28, 2020, 12:18 AM (UK time)
Update at: October 22, 2020, 6:40 PM (UK time)
Inventory override by Merri Food Hub
Supplier info: Price $5, On hand = 0, On demand = No
MFH: Price $5, On hand = On demand, On demand = Yes

Subscription info
Subscription ID: 405

image

Red flag! There is nothing in the Price estimate field in Metabase for Variant ID: #40925!
image

Product info
Product name: Campo
Product ID: #23970
Variant ID: #62890
Supplier Name: Back Alley Bakes
Supplier ID: # 3709
Last updated the same time it was created
No inventory overrides

Subscription info
Subscription ID: 396

image

Red flag! As with the other subscription, there is nothing in the Price estimate field in Metabase for Variant ID: #62890!
image

## Workaround

The workaround does not work in this scenario.

## Severity
S2

## Your Environment

* Browser name and version: Google Chrome Version 100.0.4896.88 (Official Build) (x86_64)
* Operating System and version (desktop or mobile): MacBook Pro MacOS Catalina Version 10.15.7 (19H1824)

### PR DESCRIPTION
**Pull request by georgethoppil**
#### What? Why?

Closes #9100 

When an order cycle is expired or inactive and we try to add an item to a subscription, we see that the total for the item is $NaN and the total for the order gets corrupted and also displays $NaN. This is expected since we don't have an order cycle to calculate the fees. Update the order cycle to be upcoming, it will display the estimated price now. Previously, this still displayed as $NaN

#### What should we test?
The ticket has a good instruction from Filipe [here](https://github.com/openfoodfoundation/openfoodnetwork/issues/9100#issuecomment-1114997365)

#### Release notes
Changelog Category: Technical changes
The title of the pull request will be included in the release notes.

### CODE REVIEW

File: app/controllers/admin/order_cycles_controller.rb [hunk 1]
```diff
@@ -86,6 +88,26 @@ def bulk_update
...
 
+    def bulk_update_nil_subscription_line_items_price_estimate
+      @collection.upcoming.each do |order_cycle|
+        update_nil_subscription_line_items_price_estimate(order_cycle)
+      end
+    end
+
+    def update_nil_subscription_line_items_price_estimate(order_cycle)
+      order_cycle.schedules.each do |schedule|
+        subscription = Subscription.find_by(schedule_id: schedule.id)
```
Comment (mkllnk) [inline app/controllers/admin/order_cycles_controller.rb:99]: Hm, can't there be more than one subscription per schedule?

Comment (georgethoppil) [reply app/controllers/admin/order_cycles_controller.rb]: yeah you are correct. Thanks for pointing that out

**Review by mkllnk — APPROVED** [review]
Looks good to me.

**Comment by drummer83** [comment]
Thanks @georgethoppil for this pull request! :muscle: 

First tried to reproduce the issue with master:
- Adding a product while order cycle is open works just fine
- Set close date of the order cycle with subscription to the past
- Product can't be added on subscription main page --> Error message on saving
- Product seems to be able to be added on the 'edit products' page --> 'Saved', and present on 'Review'
- BUT: Added product is not displayed on subscriptions main page

Details in Bugsnag (?): https://app.bugsnag.com/yaycode/openfoodnetwork-uk/errors/62937bf2b34daf0008bc2c26?event_id=62937bf20094a29d34b20000&i=sk&m=nw

Behavior is different than described in the issue:
Peek 2022-05-29 16-11

After staging this PR, the behavior is still exactly the same:
Peek 2022-05-29 16-34

Since I didn't manage to add products with $NaN to a subscription I cannot say if the price would be estimated correctly for future order cycles.

**Positive:** 
- I don't see any $NaN as prices (after saving/refreshing).

**Negative:** 
- I can't add any products to subscriptions with closed order cycles at all
- Even though there is a confirmation message saying 'saved' and clicking on 'review' shows the added product it is not shown anymore after refreshing the page

I couldn't see any improvement with this PR.

Asking @filipefurtad0 to double check if I have missed something. 
After deciding/fixing, issues should be created if the thing in the 'negative' list above still exist.

**Comment by filipefurtad0** [comment]
Hey @georgethoppil ,

Following the steps on that hub/sub it was possible to add a new item to the subscription - and see the `$NaN` error. -> I've noticed as well, for some other combination I could not add the Item to the sub. I did not further investigate as it seemed out of scope (and probably a rabbit hole). There's indeed some gray area here... ping @drummer83.

In any case, as seen below, **after staging the PR** and adding a new item to the subscription & setting the OC as **upcoming** (yellow) we can now check the correct prices:

Peek 2022-06-02 17-18

It's relevant to notice that if we **open** (green) the OC instead, then the `$NaN` error persists:

Peek 2022-06-02 17-14

Subscriptions is a maze... Thanks for this improvement @georgethoppil and @drummer83 for testing.

Merging.

**Pull request merged**